Frontend/src/ui/App.jsx

```diff
@@ -1,23 +1,44 @@
-import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
+import {
+  QueryClient,
+  QueryClientProvider,
+  useQuery,
+} from '@tanstack/react-query'
 import { ReactQueryDevtools } from '@tanstack/react-query-devtools'
-import React, { useCallback, useState } from 'react'
-import { AuthContext } from '../api/helper/context/auth_context'
+import React from 'react'
+import { getPermissions } from '../api/auth/get_permissions'
+import { PermissionsContext } from '../api/helper/context/permission_context'
+import { UserContext } from '../api/helper/context/user_context'
+import getMe from '../api/user/get/get_me'
+import LoadingMessage from './messages/loadingMessage'
 
 const queryClient = new QueryClient()
 export const USER_KEY = 'user'
 export const JWT_KEY = 'jwt'
 export default function App({ children }) {
-  const [user, setUserState] = useState(localStorage.getItem(USER_KEY))
-  const setUser = useCallback((user) => {
-    localStorage.setItem(USER_KEY, user)
-    localStorage.removeItem(JWT_KEY)
-    setUserState(user)
-  }, [])
+  const { isLoading, data: user } = useQuery({
+    queryKey: ['user-me'],
+    queryFn: () => getMe(),
+  })
+  const { isLoading: permissionsLoading, data: permissions } = useQuery({
+    queryKey: ['permissions-me'],
+    queryFn: () => getPermissions(),
+  })
   return (
     <QueryClientProvider client={queryClient}>
-      <AuthContext.Provider value={{ user, setUser }}>
-        {children}
-      </AuthContext.Provider>
+      <UserContext.Provider value={{ user }}>
+        <PermissionsContext.Provider
+          value={{
+            permissions,
+            canAttendCompetition: () =>
+              permissions.can_attend_competitions.scope === '*',
+            canAdminCompetition: (id) =>
+              permissions.can_administer_competitions.scope === '*' ||
+              permissions.can_administer_competitions.scope.includes(id),
+          }}
+        >
+          {isLoading || permissionsLoading ? <LoadingMessage /> : children}
```

Comment:
You can expose "loading" from the UserContext and then make a component that renders the app loading the auth if it is.

I also highly recommend caching the auth token / jwt in local storage to make the initial render nicer. I do this for competitiongroups.com.  You can use the expiresAt to determine if the JWT expired. You can also do refresh tokens with JWTs so really as long as someone has a cached JWT which is valid, then there's almost no reason to show a loading stage for user auth.